packages/fxa-auth-server/lib/routes/oauth/authorization.js

```diff
@@ -0,0 +1,244 @@
+            is: RESPONSE_TYPE_CODE,
+            then: Joi.optional(),
+            otherwise: Joi.forbidden(),
+          })
+          .when('code_challenge', {
+            is: Joi.string().required(),
+            then: Joi.required(),
+          }),
+        code_challenge: Joi.string()
+          .length(PKCE_CODE_CHALLENGE_LENGTH)
+          .when('response_type', {
+            is: RESPONSE_TYPE_CODE,
+            then: Joi.optional(),
+            otherwise: Joi.forbidden(),
+          }),
+        keys_jwe: validators.jwe.when('response_type', {
+          is: RESPONSE_TYPE_CODE,
+          then: Joi.optional(),
+          otherwise: Joi.forbidden(),
+        }),
+        acr_values: Joi.string().max(256).optional().allow(null),
+
+        resource: validators.resourceUrl.when('response_type', {
+          is: RESPONSE_TYPE_TOKEN,
+          then: Joi.optional(),
+          otherwise: Joi.forbidden(),
+        }),
+      },
+    },
+    response: {
+      schema: Joi.object()
+        .keys({
+          redirect: Joi.string(),
+          code: Joi.string(),
+          state: Joi.string(),
+          access_token: validators.accessToken,
+          token_type: Joi.string().valid('bearer'),
+          scope: Joi.string().allow(''),
+          auth_at: Joi.number(),
+          expires_in: Joi.number(),
+        })
+        .with('access_token', ['token_type', 'scope', 'auth_at', 'expires_in'])
+        .with('code', ['state', 'redirect'])
+        .without('code', ['access_token']),
+    },
+    handler: async function authorizationEndpoint(req) {
+      // Refuse to generate new codes or tokens for disabled clients.
+      if (DISABLED_CLIENTS.has(req.payload.client_id)) {
+        throw AppError.disabledClient(req.payload.client_id);
+      }
+
+      const claims = await verifyAssertion(req.payload.assertion);
+
+      const client = await oauthDB.getClient(
+        Buffer.from(req.payload.client_id, 'hex')
+      );
+      if (!client) {
+        log.debug('notFound', { id: req.payload.client_id });
+        throw AppError.unknownClient(req.payload.client_id);
+      }
+      validateClientDetails(client, req.payload, log);
+      const grant = await validateRequestedGrant(claims, client, req.payload);
+      switch (req.payload.response_type) {
+        case RESPONSE_TYPE_CODE:
+          return await generateAuthorizationCode(
+            client,
+            req.payload,
+            grant,
+            log,
+            oauthDB
+          );
+        case RESPONSE_TYPE_TOKEN:
+          return await generateImplicitGrant(client, req.payload, grant, log);
+        default:
+          // Joi validation means this should never happen.
+          log.fatal('joi.response_type', {
+            response_type: req.payload.response_type,
+          });
+          throw AppError.invalidResponseType();
+      }
+    },
+  },
+});
+
+async function generateAuthorizationCode(client, payload, grant, log, oauthDB) {
```

Comment:
I see you chose to pass `log` and `oauthDB` as arguments. This is different from how I converted `/token` where I used a closure. IDK which we should favor but we should stick to one style.

Reply:
Closures work for me.  I'll update.  👍